FSW/src/util/hkUtil.cpp

```diff
@@ -0,0 +1,184 @@
+/* hkUtil.cpp handles NS2 payload housekeeping
+ * Usage:
+ *  module functionality
+ *      system state data collection
+ *      temperature control
+ * 
+ * Modules encompassed:
+ *  houskeeping
+ *  thermal control
+ *
+ * Additional files needed for compilation:
+ *  none
+ */
+
+/* - - - - - - Includes - - - - - - */
+// C++ libraries
+
+// Other libraries
+
+// NS2 headers
+#include "../headers/hkUtil.hpp"
+
+
+/* Module Variable Definitions */
+static HousekeepingData hkData[HK_SAMPLES_TO_KEEP] = {}; // array to store housekeeping data
+static int hkIndex = 0;
+
+/* - - - - - - Module Driver Functions - - - - - - */
+
+/* - - - - - - handleHousekeeping - - - - - - *
+ * Usage:
+ *  Handles all housekeeping functions.
+ *  Samples temperature and power data, handles thermal control
+ *  
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void handleHousekeeping()
+{
+    sampleHousekeepingData();
+    setHeater();
+    hkIndex++;
+    if (hkIndex >= HK_SAMPLES_TO_KEEP) { hkIndex = 0; }
+}
+
+/* - - - - - - Module Helper Functions - - - - - - */
+
+/* - - - - - - setHeater - - - - - - *
+ * Usage:
+ *  Toggles heater based on current optics bench temperature
+ *  Turns heater on when optics temp crosses below HEATER_TEMP_LOW
+ *  Turns heater off when optics temp crosses above HEATER_TEMP_HIGH
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void setHeater()
+{
+    if (latestHkSample.opticsTemp >= HEATER_TEMP_HIGH)
+    {
+        digitalWrite(PIN_HEAT, LOW); // turn heater off
+    }
+    else if (latestHkSample.opticsTemp <= HEATER_TEMP_LOW)
+    {
+        digitalWrite(PIN_HEAT, HIGH); // turn heater on
+    }
+}
+
+/* - - - - - - sampleHousekeepingData - - - - - - *
+ * Usage:
+ *  samples temperature and power data and stores it in hkData,
+ *  checks for out-of-bounds values and logs faults if found
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void sampleHousekeepingData()
+{
+    // read thermistor voltage values
+    float opticsTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_OPTICS_THERM);
+    float analogTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_ANALOG_THERM);
+    float digitalTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_DIGITAL_THERM);
+    
+    // update housekeeping data
+    latestHkSample.opticsTemp = voltageToTemp(opticsTempVoltage); 
+    latestHkSample.analogTemp = voltageToTemp(analogTempVoltage);
+    latestHkSample.digitalTemp = voltageToTemp(digitalTempVoltage);
+    latestHkSample.analogCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_AREG_CURR);
+    latestHkSample.digitalCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_CURR);   
+    latestHkSample.digitalRegPG = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_PG);
+    hkData[hkIndex] = latestHkSample; // update data array
+
+    // Check if values are in acceptable ranges
+    // TODO: Log a fault for every case
+    if (latestHkSample.opticsTemp > OPTICS_TEMP_MAX_SAFE) { } // optics temp too high
```

Comment:
Seems like there's a redundancy between these first two if-else-if statements and the checks in `setHeater()`. I'd either just turn the heater on and off inside `sampleHousekeepingData` or I would put the fault logging in `setHeater()`.

Reply:
There is a difference between the MAX/MIN safe temperatures and the LOW/HIGH heater trigger temperatures. like this:

|MIN_SAFE| <- - - - - |HEATER_LOW(turn on)| <- -(target range)- -> |HEATER_HIGH(turn off)| - - - - - -> |MAX_SAFE|
                                                                            
setHeater checks temperature against HEATER_TEMP_LOW and HEATER_TEMP_HIGH only, and we would expect the temperature to drift a little bit outside that range while the heater is changing state, but if for some reason the temperature drifts well beyond the target range and outside of the safe range we will want to log a fault.